pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/reader/ForwardIndexReader.java

```diff
@@ -916,4 +916,83 @@ default byte[][] getBytesMV(int docId, T context) {
   default int getNumValuesMV(int docId, T context) {
     throw new UnsupportedOperationException();
   }
+
+  class ValueRange {
+    private final long _offset;
+    private final int _size;
+    // To tell if size uses bit or byte as unit, for fwd index reader reading values of fixed bits.
+    private final boolean _isSizeOfBit;
+
+    public static ValueRange newByteRange(long offset, int sizeInBytes) {
+      return new ValueRange(offset, sizeInBytes, false);
+    }
+
+    public static ValueRange newBitRange(long offset, int sizeInBits) {
+      return new ValueRange(offset, sizeInBits, true);
+    }
+
+    private ValueRange(long offset, int size, boolean isSizeOfBit) {
+      _offset = offset;
+      _size = size;
+      _isSizeOfBit = isSizeOfBit;
+    }
+
+    public long getOffset() {
+      return _offset;
+    }
+
+    public int getSize() {
+      return _size;
+    }
+
+    public boolean isSizeOfBit() {
+      return _isSizeOfBit;
+    }
+
+    @Override
+    public String toString() {
+      return "Range{" + "_offset=" + _offset + ", _size=" + _size + ", _isSizeOfBit=" + _isSizeOfBit + '}';
+    }
+  }
+
+  /**
+   * An interface that enables the caller to get byte ranges accessed while reading a given docId.
+   * This can be used to eventually allow prefetching of exact byte ranges during query processing,
+   * if the docIds to be fetched are known beforehand.
+   *
+   * This interface will be implemented by all the forward index readers.
+   */
+  interface ValueRangeProvider<T extends ForwardIndexReaderContext> {
+    /**
+     * Returns a list of {@link ValueRange} that represents all the distinct
+     * buffer byte ranges (absolute offset, sizeInBytes) that are accessed when reading the given (@param docId}
+     * @param docId to find the range for
+     * @return List of {@link ValueRange} that are accessed while reading the given docId
+     */
+    List<ValueRange> getDocIdRange(int docId, T context, @Nullable List<ValueRange> ranges);
```

Comment:
could you please help me understand why we need to return the same type (do we want to change the return type to void?)? IIUC, we want to modify the parameter `List<ValueRange> ranges` directly?

Reply:
We actually don't need to.. I had initially thought we'd only allow reading but introduced reusing the list as an after thought optimization. Changed the return type to void